### PR TITLE
Add scene validation checks: cycles detection, repeated node ids and unused nodes checks

### PR DESCRIPTION
**Pull request by WojciechBarczynski**

### CODE REVIEW

File: compositor_common/src/validators.rs [hunk 1]
```diff
@@ -98,4 +115,269 @@ impl SceneSpec {
...
+        if !unused_transforms.is_empty() {
+            return Err(SpecValidationError::UnusedNodes(unused_transforms));
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
```
Comment (wkozyra95) [inline compositor_common/src/validators.rs:243]: this will grow to much, move it to separate file

File: compositor_common/src/validators.rs [hunk 2]
```diff
@@ -98,4 +115,269 @@ impl SceneSpec {
...
+                &mut visited,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn validate_node_ids_uniqueness(
+        &self,
+        defined_node_ids: &Vec<NodeId>,
```
Comment (wkozyra95) [inline compositor_common/src/validators.rs:172]: this could be an iterator instead

File: compositor_common/src/validators.rs [hunk 3]
```diff
@@ -98,4 +115,269 @@ impl SceneSpec {
...
+    ) -> Result<(), SpecValidationError> {
+        let mut nodes_ids: HashSet<NodeId> = HashSet::new();
```
Comment (wkozyra95) [inline compositor_common/src/validators.rs:174]:
```suggestion
        let mut nodes_ids: HashSet<&NodeId> = HashSet::new();
```
clone only when returning error, but there is no need for that to just keep track of which nodes were visited

File: compositor_common/src/validators.rs [hunk 4]
```diff
@@ -98,4 +115,269 @@ impl SceneSpec {
...
+        for node_id in defined_node_ids {
+            match nodes_ids.get(node_id) {
+                Some(_) => {
+                    return Err(SpecValidationError::DuplicateNames(node_id.0.clone()));
+                }
+                None => {
+                    nodes_ids.insert(node_id.clone());
+                }
+            }
+        }
```
Comment (wkozyra95) [inline compositor_common/src/validators.rs:185]:
```suggestion
        for node_id in defined_node_ids {
             if nodes_ids.contains(node_id) {
                    return Err(SpecValidationError::DuplicateNames(node_id.0.clone()));
             }
             nodes_ids.insert(node_id);
        }
```

**Review by wkozyra95 — APPROVED** [review]

File: compositor_common/src/validators.rs (hunk 3 above)
Comment (WojciechBarczynski) [reply compositor_common/src/validators.rs]: I modified those functions' lifetimes and removed unnecessary cloning

File: compositor_common/src/validators.rs [hunk 5]
```diff
@@ -98,4 +115,125 @@ impl SceneSpec {
...
+            }
+            nodes_ids.insert(node_id);
+        }
+
+        Ok(())
+    }
+
+    fn validate_nodes_are_used(
+        &self,
+        input_nodes: &HashSet<NodeId>,
```
Comment (wkozyra95) [inline compositor_common/src/validators.rs:188]: the name here is input_nodes, but you are passing ids of registered inputs (they are not the same). Make sure to use consistent naming in other places too

But it's also not just naming, I think what we are validating here is also inconsistent. C
- You can have registered input that is not used
- You shouldn't have an input node that is not used

but this code ignores if input nodes are unused

Comment (WojciechBarczynski) [reply compositor_common/src/validators.rs]: Good point. Fixed that

**Review by wkozyra95 — APPROVED** [review]

File: compositor_common/src/validators.rs [hunk 6]
```diff
@@ -98,4 +122,134 @@ impl SceneSpec {
...
+
+        let mut visited: HashMap<&NodeId, NodeState> = HashMap::new();
+
+        fn visit<'a>(
+            node: &'a NodeId,
+            transform_nodes: &'a HashMap<&NodeId, &TransformNodeSpec>,
+            inputs: &HashSet<NodeId>,
+            visited: &mut HashMap<&'a NodeId, NodeState>,
+        ) -> Result<(), SpecValidationError> {
+            if let Some(_input) = inputs.get(node) {
```
Comment (jerzywilczek) [inline compositor_common/src/validators.rs:144]:
```suggestion
            if inputs.get(node).is_some() {
```

File: compositor_common/src/validators.rs [hunk 7]
```diff
@@ -98,4 +122,134 @@ impl SceneSpec {
...
+        &self,
+        defined_node_ids: I,
+    ) -> Result<(), SpecValidationError> {
+        let mut nodes_ids: HashSet<&NodeId> = HashSet::new();
+
+        for node_id in defined_node_ids {
+            if nodes_ids.contains(&node_id) {
+                return Err(SpecValidationError::DuplicateNames(node_id.0.clone()));
+            }
+            nodes_ids.insert(node_id);
```
Comment (jerzywilczek) [inline compositor_common/src/validators.rs:187]:
```suggestion
            if !nodes_ids.insert(node_id) {
                return Err(SpecValidationError::DuplicateNames(node_id.0.clone()));
            }
```

`insert` returns whether the insertion actually happened or the value was already in the set

File: compositor_common/src/validators.rs [hunk 8]
```diff
@@ -98,4 +122,134 @@ impl SceneSpec {
...
+
+        let mut unused_transforms: HashSet<Arc<str>> = HashSet::new();
+
+        let nodes_iter = input_nodes.keys().chain(transform_nodes.keys());
+
+        for node in nodes_iter {
+            if !visited.contains(node) {
+                unused_transforms.insert(node.0.clone());
+            }
+        }
```
Comment (jerzywilczek) [inline compositor_common/src/validators.rs:244]: You could do this using `HashSet::difference` instead

**Review by jerzywilczek — APPROVED** [review]
only cosmetic stuff

**Pull request merged**